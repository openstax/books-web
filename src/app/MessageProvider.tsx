import { shouldPolyfill } from '@formatjs/intl-pluralrules/should-polyfill';
import React from 'react';
import { IntlProvider } from 'react-intl';
import enMessages from './messages/en';

// https://formatjs.io/docs/polyfills/intl-pluralrules/#dynamic-import--capability-detection
async function polyfill(locale: 'en') {
  if (shouldPolyfill()) {
    await import('@formatjs/intl-pluralrules/polyfill');
  }

  // boolean added by the polyfill
  if ((Intl.PluralRules as (typeof Intl.PluralRules & {polyfilled?: boolean})).polyfilled) {
    switch (locale) {
      case 'en':
        await import('@formatjs/intl-pluralrules/locale-data/en');
        break;
      default:
        throw new Error('unsupported locale in messageProvider');
    }
  }
}

polyfill('en');

// tslint:disable-next-line:variable-name
const MessageProvider: React.SFC<{onError?: () => void}> = (props) =>
  <IntlProvider onError={props.onError} locale='en' messages={enMessages}>{props.children}</IntlProvider>;

export default MessageProvider;
