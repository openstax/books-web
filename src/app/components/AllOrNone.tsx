import React from 'react';
import { FormattedMessage } from 'react-intl';
import styled from 'styled-components/macro';
import { ButtonLink } from './Button';

interface Props {
  className?: string;
  onAll: () => void;
  onNone: () => void;
}

// tslint:disable-next-line:variable-name
const AllOrNone = ({className, onAll, onNone}: Props) => <div className={className}>
  <FormattedMessage id='i18n:highlighting:filters:all'>
    {(msg: Element | string) => <ButtonLink decorated onClick={onAll}>{msg}</ButtonLink>}
  </FormattedMessage>
  <span>|</span>
  <FormattedMessage id='i18n:highlighting:filters:none'>
    {(msg: Element | string) => <ButtonLink decorated onClick={onNone}>{msg}</ButtonLink>}
  </FormattedMessage>
</div>;

export default styled(AllOrNone)`
  &, ${ButtonLink} {
    font-size: 1.4rem;
  }
  height: 2rem;
  display: flex;
  flex-direction: row;
  align-items: center;

  span {
    padding: 0 1rem;
  }
`;
