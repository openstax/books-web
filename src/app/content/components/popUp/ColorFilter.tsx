import { HighlightColorEnum } from '@openstax/highlighter/dist/api';
import React from 'react';
import { FormattedMessage } from 'react-intl';
import styled from 'styled-components/macro';
import AllOrNone from '../../../components/AllOrNone';
import Checkbox from '../../../components/Checkbox';
import { textStyle } from '../../../components/Typography/base';
import { match, not } from '../../../fpUtils';
import theme from '../../../theme';
import { highlightStyles } from '../../constants';
import ColorIndicator from '../../highlights/components/ColorIndicator';
import { SummaryFilters } from '../../highlights/types';
import { filters, mobilePaddingSides } from '../../styles/PopupConstants';

// tslint:disable-next-line: variable-name
const ColorLabel = styled.span`
  flex: 1;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`;

interface Props {
  className?: string;
  styles: typeof highlightStyles;
  selectedColorFilters: Set<HighlightColorEnum>;
  colorFiltersWithContent: Set<HighlightColorEnum>;
  setSummaryFilters: (filters: Partial<SummaryFilters>) => void;
  labelKey: (label: HighlightColorEnum) => string;
}

// tslint:disable-next-line:variable-name
const ColorFilter = ({
  className,
  styles,
  selectedColorFilters,
  colorFiltersWithContent,
  setSummaryFilters,
  labelKey,
}: Props) => {

  const setSelectedColors = (colors: HighlightColorEnum[]) => {
    setSummaryFilters({colors});
  };

  const handleChange = (label: HighlightColorEnum) => {
    if (selectedColorFilters.has(label)) {
      setSelectedColors([...selectedColorFilters].filter(not(match(label))));
    } else {
      setSelectedColors([...selectedColorFilters, label]);
    }
  };

  return <div className={className} tabIndex={-1}>
    <AllOrNone
      onNone={() => setSelectedColors([])}
      onAll={() => setSelectedColors(Array.from(colorFiltersWithContent))}
    />
    {styles.map((style) => <Checkbox
      key={style.label}
      checked={selectedColorFilters.has(style.label)}
      disabled={!colorFiltersWithContent.has(style.label)}
      onChange={() => handleChange(style.label)}
    >
      <ColorIndicator style={style} size='small'/>
      <ColorLabel>
        <FormattedMessage id={labelKey(style.label)}>
          {(msg: Element | string) => msg}
        </FormattedMessage>
      </ColorLabel>
    </Checkbox>)}
  </div>;
};

export default styled(ColorFilter)`
  background: ${theme.color.white};
  display: flex;
  flex-direction: column;
  ${textStyle}
  font-size: 1.4rem;
  padding: ${filters.dropdownContent.padding.topBottom}rem ${filters.dropdownContent.padding.sides}rem;
  outline: none;
  z-index: 1;
  /* 12rem is a width of Chapter button and then we add one side padding */
  max-width: calc(100vw - ${12 + mobilePaddingSides / 2}rem);

  ${AllOrNone} {
    margin: 0.8rem 0 0.8rem 0.8rem;
  }

  ${Checkbox} {
    padding: 0.8rem;
    text-transform: capitalize;
  }

  ${ColorIndicator} {
    margin: 0 1.6rem 0 1.6rem;
  }
`;
