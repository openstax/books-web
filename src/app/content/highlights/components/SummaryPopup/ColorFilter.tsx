import React from 'react';
import styled from 'styled-components/macro';
import AllOrNone from '../../../../components/AllOrNone';
import Checkbox from '../../../../components/Checkbox';
import { textStyle } from '../../../../components/Typography/base';
import { match, not } from '../../../../fpUtils';
import { highlightStyles } from '../../constants';
import ColorIndicator from '../ColorIndicator';

interface Props {
  className?: string;
}

// tslint:disable-next-line:variable-name
const ColorFilter = ({className}: Props) => {
  const allColors = highlightStyles.map((style) => style.label);
  const [selectedColors, setSelectedColors] = React.useState<string[]>(allColors);

  return <div className={className} tabIndex={-1}>
    <AllOrNone
      onNone={() => setSelectedColors([])}
      onAll={() => setSelectedColors(allColors)}
    />
    {highlightStyles.map((style) => <Checkbox
      key={style.label}
      checked={selectedColors.includes(style.label)}
      onChange={() => selectedColors.includes(style.label)
        ? setSelectedColors(selectedColors.filter(not(match(style.label))))
        : setSelectedColors([...selectedColors, style.label])
      }
    >
      <ColorIndicator style={style} size='small'/>
      {style.label}
    </Checkbox>)}
  </div>;
};

export default styled(ColorFilter)`
  display: flex;
  flex-direction: column;
  ${textStyle}
  font-size: 1.4rem;
  padding: 0.8rem 1.6rem;
  outline: none;

  ${AllOrNone} {
    margin: 0.8rem 0 0.8rem 0.8rem;
  }

  ${Checkbox} {
    padding: 0.8rem;
  }

  ${ColorIndicator} {
    margin: 0 1.6rem 0 1.6rem;
  }
`;
