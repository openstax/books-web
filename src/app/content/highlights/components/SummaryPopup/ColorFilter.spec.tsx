import React from 'react';
import renderer from 'react-test-renderer';
import AllOrNone from '../../../../components/AllOrNone';
import Checkbox from '../../../../components/Checkbox';
import MessageProvider from '../../../../MessageProvider';
import ColorFilter from './ColorFilter';

describe('ColorFilter', () => {
  it('matches snapshot', () => {
    const component = renderer.create(<MessageProvider>
      <ColorFilter />
    </MessageProvider>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('unchecks colors', () => {
    const component = renderer.create(<MessageProvider>
      <ColorFilter />
    </MessageProvider>);

    const [box1, box2] = component.root.findAllByType(Checkbox);

    expect(box1.props.checked).toBe(true);
    expect(box2.props.checked).toBe(true);

    renderer.act(() => {
      box1.props.onChange();
    });

    expect(box1.props.checked).toBe(false);
    expect(box2.props.checked).toBe(true);
  });

  it('checks colors', () => {
    const component = renderer.create(<MessageProvider>
      <ColorFilter />
    </MessageProvider>);

    const [box1, box2] = component.root.findAllByType(Checkbox);

    expect(box1.props.checked).toBe(true);
    expect(box2.props.checked).toBe(true);

    renderer.act(() => {
      box1.props.onChange();
    });
    renderer.act(() => {
      box2.props.onChange();
    });

    expect(box1.props.checked).toBe(false);
    expect(box2.props.checked).toBe(false);

    renderer.act(() => {
      box1.props.onChange();
    });

    expect(box1.props.checked).toBe(true);
    expect(box2.props.checked).toBe(false);
  });

  it('selects none', () => {
    const component = renderer.create(<MessageProvider>
      <ColorFilter />
    </MessageProvider>);

    const [box1, box2] = component.root.findAllByType(Checkbox);
    const allOrNone = component.root.findByType(AllOrNone);

    expect(box1.props.checked).toBe(true);
    expect(box2.props.checked).toBe(true);

    renderer.act(() => {
      allOrNone.props.onNone();
    });

    expect(box1.props.checked).toBe(false);
    expect(box2.props.checked).toBe(false);
  });

  it('selects all', () => {
    const component = renderer.create(<MessageProvider>
      <ColorFilter />
    </MessageProvider>);

    const [box1, box2] = component.root.findAllByType(Checkbox);
    const allOrNone = component.root.findByType(AllOrNone);

    renderer.act(() => {
      allOrNone.props.onNone();
    });

    expect(box1.props.checked).toBe(false);
    expect(box2.props.checked).toBe(false);

    renderer.act(() => {
      allOrNone.props.onAll();
    });

    expect(box1.props.checked).toBe(true);
    expect(box2.props.checked).toBe(true);
  });
});
