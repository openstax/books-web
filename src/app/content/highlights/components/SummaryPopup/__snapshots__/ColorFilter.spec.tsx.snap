// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorFilter matches snapshot 1`] = `
<div
  className="ColorFilter-sc-1rb80m9-0 hSgVHh"
  tabIndex={-1}
>
  <div
    className="AllOrNone-sc-7m1y94-0 divvKQ"
  >
    <button
      className="Button__PlainButton-ayg7nk-2 Button__ButtonLink-ayg7nk-3 gpSxlQ"
      onClick={[Function]}
    >
      All
    </button>
    <span>
      |
    </span>
    <button
      className="Button__PlainButton-ayg7nk-2 Button__ButtonLink-ayg7nk-3 gpSxlQ"
      onClick={[Function]}
    >
      None
    </button>
  </div>
  <label
    className="Checkbox-f0dje9-2 ksczzL"
  >
    <input
      checked={true}
      disabled={false}
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="Checkbox__CustomCheckbox-f0dje9-1 gLNaiW"
    >
      <svg
        aria-hidden="true"
        className="Checkbox__CheckIcon-f0dje9-0 irsTpS StyledIconBase-sc-bdy9j4 cCvWXa"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 512 512"
      >
        <path
          d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
          fill="currentColor"
        />
      </svg>
    </span>
    <div
      className="ColorIndicator-sc-1w5zl7g-1 kBHzhz"
    >
      <svg
        aria-hidden="true"
        className="ColorIndicator__CheckIcon-sc-1w5zl7g-0 Lufbl StyledIconBase-sc-bdy9j4 cCvWXa"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 512 512"
      >
        <path
          d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
          fill="currentColor"
        />
      </svg>
    </div>
    yellow
  </label>
  <label
    className="Checkbox-f0dje9-2 ksczzL"
  >
    <input
      checked={true}
      disabled={false}
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="Checkbox__CustomCheckbox-f0dje9-1 gLNaiW"
    >
      <svg
        aria-hidden="true"
        className="Checkbox__CheckIcon-f0dje9-0 irsTpS StyledIconBase-sc-bdy9j4 cCvWXa"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 512 512"
      >
        <path
          d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
          fill="currentColor"
        />
      </svg>
    </span>
    <div
      className="ColorIndicator-sc-1w5zl7g-1 dSRHOr"
    >
      <svg
        aria-hidden="true"
        className="ColorIndicator__CheckIcon-sc-1w5zl7g-0 Lufbl StyledIconBase-sc-bdy9j4 cCvWXa"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 512 512"
      >
        <path
          d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
          fill="currentColor"
        />
      </svg>
    </div>
    green
  </label>
  <label
    className="Checkbox-f0dje9-2 dfcJwl"
  >
    <input
      checked={false}
      disabled={true}
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="Checkbox__CustomCheckbox-f0dje9-1 hiDaBp"
    >
      <svg
        aria-hidden="true"
        className="Checkbox__CheckIcon-f0dje9-0 irsTpS StyledIconBase-sc-bdy9j4 cCvWXa"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 512 512"
      >
        <path
          d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
          fill="currentColor"
        />
      </svg>
    </span>
    <div
      className="ColorIndicator-sc-1w5zl7g-1 eibKCM"
    >
      <svg
        aria-hidden="true"
        className="ColorIndicator__CheckIcon-sc-1w5zl7g-0 Lufbl StyledIconBase-sc-bdy9j4 cCvWXa"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 512 512"
      >
        <path
          d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
          fill="currentColor"
        />
      </svg>
    </div>
    blue
  </label>
  <label
    className="Checkbox-f0dje9-2 dfcJwl"
  >
    <input
      checked={false}
      disabled={true}
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="Checkbox__CustomCheckbox-f0dje9-1 hiDaBp"
    >
      <svg
        aria-hidden="true"
        className="Checkbox__CheckIcon-f0dje9-0 irsTpS StyledIconBase-sc-bdy9j4 cCvWXa"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 512 512"
      >
        <path
          d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
          fill="currentColor"
        />
      </svg>
    </span>
    <div
      className="ColorIndicator-sc-1w5zl7g-1 ekjDWZ"
    >
      <svg
        aria-hidden="true"
        className="ColorIndicator__CheckIcon-sc-1w5zl7g-0 Lufbl StyledIconBase-sc-bdy9j4 cCvWXa"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 512 512"
      >
        <path
          d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
          fill="currentColor"
        />
      </svg>
    </div>
    purple
  </label>
  <label
    className="Checkbox-f0dje9-2 ksczzL"
  >
    <input
      checked={true}
      disabled={false}
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="Checkbox__CustomCheckbox-f0dje9-1 gLNaiW"
    >
      <svg
        aria-hidden="true"
        className="Checkbox__CheckIcon-f0dje9-0 irsTpS StyledIconBase-sc-bdy9j4 cCvWXa"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 512 512"
      >
        <path
          d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
          fill="currentColor"
        />
      </svg>
    </span>
    <div
      className="ColorIndicator-sc-1w5zl7g-1 hFYYXx"
    >
      <svg
        aria-hidden="true"
        className="ColorIndicator__CheckIcon-sc-1w5zl7g-0 Lufbl StyledIconBase-sc-bdy9j4 cCvWXa"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 512 512"
      >
        <path
          d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
          fill="currentColor"
        />
      </svg>
    </div>
    pink
  </label>
</div>
`;
